arch/sim/src/sim/up_ucontext.c

```diff
@@ -0,0 +1,283 @@
+
+/****************************************************************************
+ * Name: task_exit_point
+ *
+ * Description:
+ *   This function is used to release the allocated stack memory for a task
+ *   on it's exit path.
+ *
+ * Assumptions:
+ *   This function does not return.
+ *
+ ****************************************************************************/
+
+static void task_exit_point(void)
+{
+  for (; ; )
+    {
+      int cpu_id = 0;
+
+#ifdef CONFIG_SMP
+      cpu_id = up_cpu_index();
+#endif
+      if (g_task_exit_details[cpu_id].ucontext_sp)
+        {
+          /* Free stack memory used by the ucontext structure */
+
+          free(g_task_exit_details[cpu_id].current_ucontext);
+          free(g_task_exit_details[cpu_id].ucontext_sp);
+
+          /* Activate the parent ucontext */
+
+          setcontext(g_task_exit_details[cpu_id].parent_ucontext);
+        }
+    }
+}
+
+/****************************************************************************
+ * Name: up_setup_exit_context
+ *
+ * Description:
+ *   Creates new exit context for tasks. In a SMP configuration every CPU
+ *   has a different exit context.
+ *
+ ****************************************************************************/
+
+static void up_setup_exit_context(int cpu)
+{
+  ucontext_t *exit_context = &g_task_exit_context[cpu];
+  int ret = getcontext(exit_context);
+  ASSERT(ret >= 0);
+
+  uint8_t *sp = &g_task_exit_context_stack[cpu][0];
+
+  uint64_t align_offset = STACK_ALIGNMENT_BYTES - ((uint64_t)sp %
+      STACK_ALIGNMENT_BYTES);
+
+  uint8_t *aligned_stack = sp + align_offset;
+  exit_context->uc_stack.ss_sp   = aligned_stack;
+  exit_context->uc_stack.ss_size = MIN_UCONTEXT_STACK_LENGTH -
+    align_offset;
+  exit_context->uc_link          = &g_uctx_main;
+
+  makecontext(exit_context, task_exit_point, 0);
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: up_create_context
+ *
+ * Description:
+ *   Creates a new ucontext structure and initialize it.
+ *
+ * Input Parameters:
+ *   ucontext_sp   - buffer where we store the ucontext stack pointer
+ *   prev_ucontext - the parent ucontext used%i to return to when the
+ *    execution of the current context ends.
+ *   entry_point   - the entry point of the new context
+ *
+ * Return Value:
+ *   A pointer to the ucontext structure or NULL in case something went
+ *   wrong.
+ *
+ ****************************************************************************/
+
+void *up_create_context(void **ucontext_sp, void *prev_ucontext,
+    void (*entry_point)(void))
+{
+  ucontext_t *new_context = calloc(1, sizeof(ucontext_t));
+  ASSERT(new_context != NULL);
+
+  int ret = getcontext(new_context);
+  ASSERT(ret >= 0);
+
+  uint8_t *sp = calloc(1, MIN_UCONTEXT_STACK_LENGTH);
```

Comment:
we need use the stack allocated by NuttX's kernel.

Reply:
In that case we will need to re-adjust the stack for every thread using a defconfig because `ucontext` needs ~128K of stack(the same applies for apps which leave in a different repo). I wanted to avoid this, but if this is ok with you I will go with your recommendation.

Reply:
We don't need change for most apps, modify one line is enough at nuttx/Kconfig:
```
config DEFAULT_TASK_STACKSIZE
	int "The default stack size for tasks"
	default 8192 if ARCH_SIM
	default 2048
	---help---
		The default stack size for tasks.
```
@yamt add DEFAULT_TASK_STACKSIZE config several months ago for this particular needing.

Reply:
Right, but this is not for all the configs. For example for the `ostest` config there is another macro hidden in a header file CONFIG_TESTING_OSTEST_STACKSIZE. It should be useful to have a check if you configured something that has the stack size less than the default value: `DEFAULT_TASK_STACKSIZE`

Reply:
Yes, to keep the backward compatibility, the stack size config for some application can't change to DEFAULT_TASK_STACKSIZE. We have to set it in defconfig.

Reply:
Hi @sebastianene07 do you have some update? I'm not hurry you, but this is a nice feature to have! :-)

Reply:
Hello @acassis thanks for your interest ! Unfortunately I hit some problems on the road with this patch because it doesn't seem to work well with SMP. I'm still investigating this.